rust/template/types/closure.rs

```diff
@@ -0,0 +1,412 @@
+
+#[cfg(feature = "flatbuf")]
+use crate::flatbuf::fb;
+
+use ::differential_datalog::record::FromRecord;
+use ::differential_datalog::record::IntoRecord;
+use ::differential_datalog::record::Mutator;
+use ::differential_datalog::record::Record;
+
+use ::abomonation::Abomonation;
+
+/* DDlog's equivalent of Rust's `Fn` trait.  This is necessary, as Rust does not allow manual
+ * implementations of `Fn` trait (until `unboxed_closures` and `fn_traits` features are
+ * stabilized).  Otherwise, we would just derive `Fn` and add methods for comparison and hashing.
+ */
+pub trait Closure<Args, Output> {
+    fn call(&self, args: Args) -> Output;
+    /* Returns pointers to function and captured arguments, for use in comparison methods. */
+    fn internals(&self) -> (usize, usize);
+    fn clone_dyn(&self) -> Box<dyn Closure<Args, Output>>;
+    fn eq_dyn(&self, other: &dyn Closure<Args, Output>) -> bool;
+    fn cmp_dyn(&self, other: &dyn Closure<Args, Output>) -> Ordering;
+    fn hash_dyn(&self, state: &mut dyn Hasher);
+    fn into_record_dyn(&self) -> Record;
+    fn fmt_debug_dyn(&self, f: &mut Formatter) -> std::fmt::Result;
+    fn fmt_display_dyn(&self, f: &mut Formatter) -> std::fmt::Result;
+    fn serialize_dyn(&self) -> &dyn erased_serde::Serialize;
+}
+
+#[derive(Clone)]
+pub struct ClosureImpl<Args, Output, Captured: Val> {
+    pub description: &'static str,
+    pub captured: Captured,
+    pub f: fn(args: Args, captured: &Captured) -> Output,
+}
+
+impl<Args, Output, Captured: Debug + Val> Serialize for ClosureImpl<Args, Output, Captured> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&format!(
+            "<closure: {}, captured_args: {:?}>",
+            self.description, self.captured
+        ))
+    }
+}
+
+/* Rust forces 'static trait bound on `Args` and `Output`, as the borrow checker is not smart
+ * enough to realize that they are only used as arguments to `f`.
+ */
+impl<Args: Clone + 'static, Output: Clone + 'static, Captured: Debug + Val> Closure<Args, Output>
+    for ClosureImpl<Args, Output, Captured>
+{
+    fn call(&self, args: Args) -> Output {
+        (self.f)(args, &self.captured)
+    }
+
+    fn clone_dyn(&self) -> Box<dyn Closure<Args, Output>> {
+        Box::new((*self).clone()) as Box<dyn Closure<Args, Output>>
+    }
+
+    fn internals(&self) -> (usize, usize) {
+        (
+            self.f as *const (fn(Args, &Captured) -> Output) as usize,
+            &self.captured as *const Captured as usize,
+        )
+    }
+
+    fn eq_dyn(&self, other: &dyn Closure<Args, Output>) -> bool {
+        /* Compare function pointers.  If equal, it is safe to compare captured variables. */
+        let (other_f, other_captured) = other.internals();
+        if (other_f == (self.f as *const (fn(Args, &Captured) -> Output) as usize)) {
+            unsafe { *(other_captured as *const Captured) == self.captured }
```

Comment:
So you compare pointer values?
Isn't there a safe way to do this?

Reply:
I wonder whether this will cause problems in a distributed settings. I think that we have an invariant so far that the ordering/hash/equality of values on different machines is the same.

Reply:
> Isn't there a safe way to do this?
Not if pointers can point to functions of different types, which is the case here, because even closures of the same type can differ in the type of their captured values.

> I wonder whether this will cause problems in a distributed settings. I think that we have an invariant so far that the ordering/hash/equality of values on different machines is the same.

It's not a problem for closures, as closures cannot be sent across the network atm (we'd have to parse and compile DDlog code on the fly).  But this may be a problem, e.g., for interned values whose order will definitely change as you send them across the network. #663 would address this problem.